### ISSUE TITLE
"wrong type argument listp key" when calling "julia-repl" 

### ISSUE THREAD

**Issue by orialb** [tpapp/julia-repl#38]
Hi,
I installed the package via spacemacs julia layer so I'm not sure if the issue is related specifically to spacemacs or not. If this is the wrong place to report an issue when using the spacemacs layer let me know.

after installing, when running julia-repl command I got  "wrong type argument listp key" error.
I found a workaround so just wanted to report this in case anyone else encounters it.

The error was caused by the following function in julia-repl.el:
```lisp
(defun julia-repl--default-executable-key ()
  "Return the default executable key."
  (let ((key (caar julia-repl-executable-records)))
    (assert key "Could not find any key in JULIA-REPL-EXECUTABLE-RECORDS.")
    key))
```
the problem seems to be with the assert statement. My workaround was to comment out the assertion and then everything worked just fine.
```lisp 
(let ((key (caar julia-repl-executable-records)))
    ;;(assert key "Could not find any key in JULIA-REPL-EXECUTABLE-RECORDS.")
    key))
```
not sure why is it causing problems.

1. Emacs version: GNU emacs 25.3.1 (with spacemacs develop branch)
2. Julia version: 0.6.4
3. version of julia-repl: from spacemacs layer julia
4. self-contained steps to replicate the issue:
- install julia spacemacs layer
- run julia-repl command

**Comment by tpapp** [comment]
Thanks for the debug information. I will try to figure out what's wrong, but need a bit of help. Can you replace the function body with

```emacs-lisp
(defun julia-repl--default-executable-key ()
  "Return the default executable key."
  (message "DEBUG executable records: %s" julia-repl-executable-records) 
  (let ((key (caar julia-repl-executable-records)))
    (message "DEBUG key: %s" key)
    (assert key "Could not find any key in JULIA-REPL-EXECUTABLE-RECORDS.")
    key))
```
and report back the output (it will be in the `*Message*` buffer)?

**Comment by orialb** [comment]
This is what I get:
DEBUG executable records: ((default julia))
DEBUG key: default
assert: Wrong type argument: listp, key

**Comment by tpapp** [comment]
Weird. What does `M-: (assert 'default)` do? (ie evaluate `(assert 'default)` as elisp code).

It should simply eval to `nil`. Your version seems to throw an error.

**Comment by orialb** [comment]
`(assert 'default)` does evaluate to nil. 
but if I run `(let ((key 'default)) (assert key "hmm"))' I get the wrong type argument error,
while '(let ((key 'default)) key)' just evals to default.
So it seems that the problem is somehow with the assert inside the let statement?

**Comment by tpapp** [comment]
@orialb: I set up spacemacs and it seems it only has `cl-assert`. This should be fixed in latest `master`, please try it out and report back (and reopen the issue if it didn't).

**Comment by orialb** [comment]
`cl-assert` still doesn't work. I guess we can keep this closed in case I'm the only one encountering this issue as I'm ok with the current workaround.

**Comment by orialb** [comment]
Ok, I will open up this issue with spacemacs and update if there are any results.

**Comment by orialb** [comment]
the issue does not seem to reproduce for other spacemacs users.
seems something got messed up with my setup which causes `(assert var string)` (also with cl-assert) , although `(assert var)` is working. 
That is very weird, but since it is not reproducing for others I guess we can close this issue.

**Comment by bmag** [comment]
Saw the issue in Spacemacs, just wanted to drop a few suggestions:

- `(assert key "some string")` is wrong, at least in Emacs 26.1. Should be `(assert key nil "some string")` - check out the macro's documentation.

- `cl-assert` is from `cl-macs.el`, so you might want to require `cl-lib` (which requires `cl-macs` and `cl-*` stuff). Although, usually some other package already requires `cl-lib` (Spacemacs does too).

- are you using different versions of Emacs with the same elpa directory? If yes, be aware that byte-compiled files (.elc) are not 100% compatible between different Emacs major versions and could cause errors, especially with `cl-*` stuff I've seen it a few times. If that is the case, deleting the file `julia-repl.elc` (or just moving it somewhere else) would fix the problem.

**Comment by orialb** [comment]
thanks for the input!
- `(assert key nil "some string")` works for me (I'm on emacs 25.3.1 according to emacs-version). 
updated the assert in julia-repl.el and now everything works.

- from your other comments I understand that cl-assert should have worked for me as well? I don't think I'm using different versions of Emacs. I was on spacemacs master before and in order to switch to develop I deleted the .emacs.d dir, cloned spacemacs git repo and then checked out the develop branch (I kept my previous .spacemacs file though). can it be that this left some leftovers?

**Comment by tpapp** [comment]
@bmag: thanks for the help.

@orialb: I think I fix this in branch tp/fix-assert, in PR #39. please kindly test and let me know.

**Comment by bmag** [comment]
@orialb if `assert` works then also `cl-assert` works. If you don't think you're using different versions of Emacs then probably you aren't, so the byte-compilation part is irrelevant. The Spacemacs changes shouldn't have left any meaningful leftovers, no need to worry about it.

**Comment by orialb** [comment]
@bmag thanks.
Maybe one more question since you seem to be familiar with spacemacs. I want to test the fix suggested in the tp/fix-assert, but I guess I should install it via the julia layer to test it properly. 
Do you have a clue what would be the correct way of forcing julia layer to use this specific version of the package?
What I already tried: 
1.  cloning the tp/fix-assert branch to my .emacs.d/private/local
2. placing `(julia-repl :location local)` in julia/packages.el file (instead of :location elpa) .
3. restarting spacemacs
This doesn't seem to work and gives "File mode specification error: (void-function julia-repl-mode)" when opening a julia file. 
If I use `(julia-repl :location (recipe :fetcher github :repo "tpapp/julia-repl"))` this installs fine (but from julia-repl master).

**Comment by bmag** [comment]
@orialb use the recipe approach. You can even specify a branch:
```elisp
(julia-repl :location (recipe :fetcher github :repo "tpapp/julia-repl" :branch "branch-name"))
```

**Comment by orialb** [comment]
@bmag cool that works, thanks again :)
@tpapp  happy to confirm that the version in tp/fix-assert is working properly for me.

**Comment by tpapp** [comment]
@orialb: thanks for testing, merged #39 into `master`. will make a minor version bump release, closing the issue.

@bmag: thanks for your help with this, I really appreciate it.
